rse-bash-advanced.Rmd

```diff
@@ -702,6 +702,181 @@ $ find . -name "*.bak"
+
+```shell
+$ set
+```
+
+```text
+COMPUTERNAME=TURING
+HOME=/home/amira
+HOMEDRIVE=C:
+HOSTNAME=TURING
+HOSTTYPE=i686
+NUMBER_OF_PROCESSORS=4
+OS=Windows_NT
+PATH=/Users/amira/bin:/usr/local/git/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin
+PWD=/home/amira
+UID=1000
+USERNAME=amira
+...
+```
+
+There are quite a few—many more than are shown here.
+And yes,
+using `set` to *show* things might seem a little strange,
+even for Unix,
+but if we don't give it any arguments,
+it might as well show us things we *could* set.
+
+Every variable has a name.
+By convention, variables that are always present are given upper-case names.
+All shell variables' values are strings, even those (like `UID`) that look like numbers.
+It's up to programs to convert these strings to other types when necessary.
+For example, if a program wanted to find out how many processors the computer had,
+it would convert the value of the `NUMBER_OF_PROCESSORS` variable from a string to an integer.
+
+Similarly, some variables (like `PATH`) store lists of values.
+In this case, the convention is to use a colon ':' as a separator.
+If a program wants the individual elements of such a list,
+it's the program's responsibility to split the variable's string value into pieces.
+
+Let's have a closer look at `PATH`.
+Its value defines the shell's [search path][search-path],
+i.e., the list of directories that the shell looks in for runnable programs
+when we type in a program name without specifying what directory it is in.
+
+For example,
+when we type a command like `analyze`,
+the shell needs to decide whether to run `./analyze` or `/bin/analyze`.
+The rule it uses is simple:
+the shell checks each directory in the `PATH` variable in turn,
+looking for a program with the requested name in that directory.
+As soon as it finds a match, it stops searching and runs the program.
+
+To show how this works,
+here are the components of `PATH` listed one per line:
+
+```shell
+/Users/amira/bin
+/usr/local/git/bin
+/usr/bin
+/bin
+/usr/sbin
+/sbin
+/usr/local/bin
+```
+
+On our computer,
+there are actually three programs called `analyze`
+in three different directories:
+`/bin/analyze`,
+`/usr/local/bin/analyze`,
+and `/users/amira/analyze`.
+Since the shell searches the directories in the order they're listed in `PATH`,
+it finds `/bin/analyze` first and runs that.
+Notice that it will *never* find the program `/users/amira/analyze`
+unless we type in the full path to the program,
+since the directory `/users/amira` isn't in `PATH`.
+
+Let's show the value of the variable `HOME`:
+
+```shell
+$ echo HOME
+```
+
+```text
+HOME
+```
+
+That just prints "HOME", which isn't what we wanted
+(though it is what we actually asked for).
+Let's try this instead:
+
+```shell
+$ echo $HOME
+```
+
+```text
+/home/amira
+```
+
+The dollar sign tells the shell that we want the value of the variable
+rather than its name.
```

Comment:
Mm - neither `HOME` nor `$HOME` print anything on their own - that's the `echo`.

Reply:
Maybe we can say "rather than echoing the string `HOME`"? My point is just that we are not printing the name of a variable, we're printing a string that is independent of the potential existence of any variables.